### DISCUSSION TITLE
Tasmota SR04 sensor driver serial mode (UART) do not trigger data (A02YYUW )

### DISCUSSION THREAD

**Discussion by DaFcz** [arendst/Tasmota#22151] [Q&A] [answered]
Some  time ago I have purchased an A02YYUW - Mini Ultrasonic Sensor (it is similar to SR04 sensor module) which uses a serial UART command control with 4  byte output (Header,DATA_H,DATA_L,SUM). I have understand that Tasmota SR04 module supports both Trig/Echo and UART serial devices, but I can't to read any data from the A02YYUW when using the SR04 module.

If I connect the sensor Tx output to GPIO3 and Rx to GPIO1(of WeMos D1 Mini) and use the Tasmota module configuration  Generic(18) configured for RX GPIO3 = SerialRx / GPIO1 = SerialTx and send a "SerialSend4 1" command from the console, I can see one measurement in the console:

16:32:11.001 CMD: SerialSend4 1
16:32:11.005 RSL: RESULT = {"SerialSend":"Done"}
16:32:11.419 RSL: RESULT = {"SerialReceived":"FF05D7DB"}

When I set the module configuration so that RX GPIO3 = SR04 Ech/RX and TX GPIO1 = SR04 Tri/TX then after restart in console I can see once:

00:00:01.006 RSL: RESULT = {"SerialReceived":"FF05DADEFF05DADE"}
00:00:02.156 SR4-1: Mode 0
00:00:02.359 RSL: RESULT = {"SerialReceived":"00"}

It means SR04 driver does not recognize the sensor?
The sensor is "UART Controlled" - the sensor needs triggering signal to perform measurement. Is this mode supported by SR04 driver?

I'm using tasmota-sensors firmware version 14.2.0(release-sensors)-2_7_7
If anyone can suggest what I'm doing wrong I would be very grateful.

**Comment by DaFcz** [comment, marked as answer]
I have found solution. there is in "xsns_22_sr04.ino" hardcoded baudrate to 9600
if (SR04[i].sonar_serial && SR04[i].sonar_serial->begin(9600))
But the sensor is using 115200, when it is changed, it is working as expected.

00:00:01.006 RSL: RESULT = {"SerialReceived":"FF05E0E4FF05DEE2"}
00:00:01.947 SNS: Hardware Serial
00:00:01.948 SR4-1: Mode 3
...
19:55:16.285 RSL: sensors = {"sn":{"Time":"2024-09-16T19:55:16","SR04":{"Distance":150.4}},"ver":1}

Thank you for your code and Tasmota

**Comment by JohnSmithCz** [comment]
Hi,
I have a similar problem with A02YYUW UART Version. When entering the configuration TX GPIO1-SR04 Tri/TX and RX GPIO3 SR04 Ech/RX it write to the console (abbreviated):

21:52:43.561 MQT: tele/RESULT = {"SerialReceived":"FFEDFFFF6DFFFFFF75FF7FDBFEDFF77FFDDFEEFDF"}

Can you tell me where I can find the file "xsns_22_sr04.ino" and what should I edit? I'm using Wemos D1 mini with Tasmota 15.1.0 Sensors firmware. I'm a beginner, can you please describe in detail, thank you very much.

(I also tried not connecting the TX GPIO1 and in this case it doesn't say anything.)

**Comment by JohnSmithCz** [comment]
I tried connecting the RX and TX pins on the Wemos D1 to pins D1 and D2 (GPIO5 and GPIO4) instead and this combination works.
SR04 RX - D2 GPIO4 (TX) and 
SR04 TX - D1 GPIO5 (RX).
